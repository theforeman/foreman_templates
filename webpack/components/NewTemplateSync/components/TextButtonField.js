import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import { FormControl, InputGroup, Button, Checkbox } from 'patternfly-react';

import TextField from 'foremanReact/components/common/forms/TextField';
import CommonForm from 'foremanReact/components/common/forms/CommonForm';

const renderField = ({
  input,
  label,
  className,
  inputClassName,
  fieldRequired,
  disabled,
  blank,
  item,
  fieldSelector,
  meta: { error, touched },
  buttonAttrs: { buttonText = "Action", buttonAction },
}) => (
  <CommonForm
    label={label}
    className={className}
    inputClassName={inputClassName}
    required={fieldRequired}
    error={error}
    touched={touched}
  >
    <InputGroup>
      <FieldType item={item} fieldSelector={fieldSelector} input={input} disabled={disabled} blank={blank} />
      <InputGroup.Button>
        <Button onClick={buttonAction} disabled={disabled}>{ buttonText }</Button>
      </InputGroup.Button>
    </InputGroup>
  </CommonForm>
)

const TextButtonField = ({
  item,
  label,
  name,
  className = '',
  inputClassName = 'col-md-6',
  blank = { label: "Choose one...", value: "" },
  buttonAttrs,
  fieldSelector,
  validate,
  disabled = false,
  fieldRequired = false
}) => (
  <Field name={name}
         label={label}
         type={fieldSelector(item)}
         fieldSelector={fieldSelector}
         component={renderField}
         buttonAttrs={buttonAttrs}
         blank={blank}
         item={item}
         disabled={disabled}
         validate={item.validate}
         fieldRequired={fieldRequired}>
  </Field>
);

const FieldType = ({ item, fieldSelector, input, disabled, blank }) => {
  if (!fieldSelector) {
    return (<InputField input={input} disabled={disabled} />);
  }

  switch (fieldSelector(item)) {
    case "text":
      return <InputField input={input} disabled={disabled} />;
    case "select":
      return <SelectField input={input} blank={blank} item={item} disabled={disabled} />;
    case "checkbox":
      return <CheckboxField input={input} item={item} disabled={disabled} />;
    default:
      throw new Error(`Unknown field type ${fieldSelector(item)} for ${item}`);
  }
}

const InputField = ({ input, disabled }) =>
  <FormControl { ...input } type="text" disabled={disabled}></FormControl>

const SelectField = ({ input, blank, item, disabled }) =>
  <FormControl { ...input } componentClass="select" disabled={disabled}>
    <BlankOption blank={blank} />
    { item.selection.map((opt) => <option key={opt.value} value={opt.value}>{opt.label}</option>) }
  </FormControl>

const CheckboxField = ({ input, item, disabled }) => {
  return <Checkbox { ...input } disabled={disabled}></Checkbox>
}

const BlankOption = ({ blank }) => {
  if (Object.keys(blank).length === 0) {
    return null;
  } else {
    return <option key={blank.value} value={blank.value}>{blank.label}</option>;
  }
}

export default TextButtonField;
