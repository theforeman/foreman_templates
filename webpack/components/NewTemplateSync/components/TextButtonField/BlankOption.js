import React from 'react';
import PropTypes from 'prop-types';

const BlankOption = ({ blank }) => {
  if (Object.keys(blank).length === 0) {
    return null;
  }
  return (
    <option key={blank.value} value={blank.value}>
      {blank.label}
    </option>
  );
};

BlankOption.propTypes = {
  blank: PropTypes.object.isRequired,
};

export default BlankOption;
