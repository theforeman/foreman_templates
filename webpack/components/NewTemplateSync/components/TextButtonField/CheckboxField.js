import React from 'react';
import { Checkbox } from 'patternfly-react';
import PropTypes from 'prop-types';

const CheckboxField = ({ input, item, disabled }) => (
  <Checkbox {...input} disabled={disabled} />
);

CheckboxField.propTypes = {
  input: PropTypes.object.isRequired,
  item: PropTypes.object.isRequired,
  disabled: PropTypes.bool.isRequired,
};

export default CheckboxField;
