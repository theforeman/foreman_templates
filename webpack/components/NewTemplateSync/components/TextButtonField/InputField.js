import React from 'react';
import { FormControl } from 'patternfly-react';
import PropTypes from 'prop-types';

const InputField = ({ input, disabled }) => (
  <FormControl {...input} type="text" disabled={disabled} />
);

InputField.propTypes = {
  input: PropTypes.object.isRequired,
  disabled: PropTypes.bool.isRequired,
};

export default InputField;
