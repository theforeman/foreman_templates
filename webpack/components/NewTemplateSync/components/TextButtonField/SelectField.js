import React from 'react';
import PropTypes from 'prop-types';
import { FormControl } from 'patternfly-react';
import BlankOption from './BlankOption';

const SelectField = ({ input, blank, item, disabled }) => (
  <FormControl {...input} componentClass="select" disabled={disabled}>
    <BlankOption blank={blank} />
    {item.selection.map(opt => (
      <option key={opt.value} value={opt.value}>
        {opt.label}
      </option>
    ))}
  </FormControl>
);

SelectField.propTypes = {
  input: PropTypes.object.isRequired,
  blank: PropTypes.object.isRequired,
  item: PropTypes.object.isRequired,
  disabled: PropTypes.bool.isRequired,
};

export default SelectField;
