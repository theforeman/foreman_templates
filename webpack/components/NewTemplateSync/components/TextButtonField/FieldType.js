import React from 'react';
import PropTypes from 'prop-types';

import InputField from './InputField';
import SelectField from './SelectField';
import CheckboxField from './CheckboxField';

const FieldType = ({ item, fieldSelector, input, disabled, blank }) => {
  if (!fieldSelector) {
    return <InputField input={input} disabled={disabled} />;
  }

  switch (fieldSelector(item)) {
    case 'text':
      return <InputField input={input} disabled={disabled} />;
    case 'select':
      return (
        <SelectField
          input={input}
          blank={blank}
          item={item}
          disabled={disabled}
        />
      );
    case 'checkbox':
      return <CheckboxField input={input} item={item} disabled={disabled} />;
    default:
      throw new Error(`Unknown field type ${fieldSelector(item)} for ${item}`);
  }
};

FieldType.propTypes = {
  item: PropTypes.object.isRequired,
  fieldSelector: PropTypes.func,
  input: PropTypes.object.isRequired,
  disabled: PropTypes.bool,
  blank: PropTypes.bool,
};

FieldType.defaultProps = {
  fieldSelector: null,
  disabled: false,
  blank: false,
};

export default FieldType;
