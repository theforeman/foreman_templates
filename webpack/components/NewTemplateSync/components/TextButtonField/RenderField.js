import React from 'react';
import CommonForm from 'foremanReact/components/common/forms/CommonForm';
import { InputGroup, Button } from 'patternfly-react';
import PropTypes from 'prop-types';

import FieldType from './FieldType';

const RenderField = ({
  input,
  label,
  className,
  inputClassName,
  fieldRequired,
  disabled,
  blank,
  item,
  fieldSelector,
  tooltipHelp,
  meta: { error, touched },
  buttonAttrs: { buttonText, buttonAction },
}) => (
  <CommonForm
    label={label}
    className={className}
    inputClassName={inputClassName}
    required={fieldRequired}
    error={error}
    touched={touched}
    tooltipHelp={tooltipHelp}
  >
    <InputGroup>
      <FieldType
        item={item}
        fieldSelector={fieldSelector}
        input={input}
        disabled={disabled}
        blank={blank}
      />
      <InputGroup.Button className="left-padded">
        <Button onClick={buttonAction} disabled={disabled}>
          {buttonText}
        </Button>
      </InputGroup.Button>
    </InputGroup>
  </CommonForm>
);

RenderField.propTypes = {
  input: PropTypes.object.isRequired,
  label: PropTypes.string.isRequired,
  className: PropTypes.string,
  inputClassName: PropTypes.string,
  fieldRequired: PropTypes.bool,
  disabled: PropTypes.bool.isRequired,
  blank: PropTypes.object,
  item: PropTypes.object.isRequired,
  fieldSelector: PropTypes.func,
  tooltipHelp: PropTypes.node,
  meta: PropTypes.object,
  buttonAttrs: PropTypes.object,
};

RenderField.defaultProps = {
  className: '',
  inputClassName: 'col-md-4',
  fieldRequired: false,
  blank: {},
  fieldSelector: null,
  tooltipHelp: null,
  meta: {},
  buttonAttrs: {},
};

export default RenderField;
