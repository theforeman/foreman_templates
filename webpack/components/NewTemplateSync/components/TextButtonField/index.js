import React from 'react';
import { get } from 'lodash';
import { Field as FormikField } from 'formik';
import PropTypes from 'prop-types';

import RenderField from './RenderField';

const TextButtonField = ({
  item = {},
  label,
  name,
  className,
  inputClassName,
  blank,
  buttonText,
  buttonAction,
  fieldSelector,
  disabled,
  fieldRequired,
  tooltipHelp,
}) => (
  <FormikField
    name={name}
    render={({ field, form }) => (
      <RenderField
        label={label}
        fieldSelector={fieldSelector}
        tooltipHelp={tooltipHelp}
        buttonAttrs={{
          buttonText,
          buttonAction: () => buttonAction(form.setFieldValue),
        }}
        blank={blank}
        item={item}
        disabled={disabled}
        fieldRequired={fieldRequired}
        meta={{
          touched: get(form.touched, name),
          error: get(form.errors, name),
        }}
        input={field}
      />
    )}
  />
);

TextButtonField.propTypes = {
  item: PropTypes.object,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  inputClassName: PropTypes.string,
  blank: PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.string,
  }),
  buttonText: PropTypes.node.isRequired,
  buttonAction: PropTypes.func.isRequired,
  fieldSelector: PropTypes.func,
  disabled: PropTypes.bool,
  fieldRequired: PropTypes.bool,
  tooltipHelp: PropTypes.node,
};

TextButtonField.defaultProps = {
  blank: { label: 'Choose one...', value: '' },
  className: '',
  inputClassName: 'col-md-6',
  disabled: false,
  fieldRequired: false,
  tooltipHelp: null,
  fieldSelector: () => 'text',
  item: {},
};

export default TextButtonField;
