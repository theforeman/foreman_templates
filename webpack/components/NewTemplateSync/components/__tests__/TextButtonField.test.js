import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import TextButtonField from '../TextButtonField';

const textItem = {
  type: 'text',
};

const selectItem = {
  type: 'select',
  selection: [
    { label: 'A', value: 'a' },
    { label: 'B', value: 'b' },
    { label: 'C', value: 'c' },
  ],
};

const checkboxItem = {
  type: 'checkbox',
};

const blank = { label: 'Bare Metal', value: 'bareMetal' };

const fieldSelector = item => item.type;
const buttonAttrs = { buttonText: 'Button', buttonAction: () => {} };

const fixtures = {
  'should render text item': {
    name: 'Text field',
    item: textItem,
    fieldSelector,
    ...buttonAttrs,
    disabled: false,
    label: 'Text',
    blank: {},
  },
  'should render checkbox item': {
    name: 'Checkbox field',
    item: checkboxItem,
    fieldSelector,
    ...buttonAttrs,
    disabled: false,
    label: 'Checkbox',
    blank: {},
  },
  'should render select item with custom blank opt': {
    name: 'select field',
    item: selectItem,
    fieldSelector,
    ...buttonAttrs,
    label: 'Select',
    disabled: false,
    blank,
  },
  'should render text field without field selector': {
    name: 'no selector',
    label: 'Text field',
    ...buttonAttrs,
    disabled: false,
    blank: {},
  },
};

describe('TextButtonField', () =>
  testComponentSnapshotsWithFixtures(TextButtonField, fixtures));
