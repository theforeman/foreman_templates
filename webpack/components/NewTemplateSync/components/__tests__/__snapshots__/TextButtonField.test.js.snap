// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextButtonField should render checkbox item 1`] = `
<FormikConnect(FieldInner)
  name="Checkbox field"
  render={[Function]}
/>
`;

exports[`TextButtonField should render select item with custom blank opt 1`] = `
<FormikConnect(FieldInner)
  name="select field"
  render={[Function]}
/>
`;

exports[`TextButtonField should render text field without field selector 1`] = `
<FormikConnect(FieldInner)
  name="no selector"
  render={[Function]}
/>
`;

exports[`TextButtonField should render text item 1`] = `
<FormikConnect(FieldInner)
  name="Text field"
  render={[Function]}
/>
`;
